Code/Framework/AzCore/AzCore/DOM/DomBackend.cpp

```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright (c) Contributors to the Open 3D Engine Project.
+ * For complete copyright and license terms please see the LICENSE at the root of this distribution.
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+ *
+ */
+
+#include <AzCore/DOM/DomBackend.h>
+
+#include <AzCore/Utils/Utils.h>
+#include <AzCore/IO/ByteContainerStream.h>
+#include <AzCore/DOM/DomBackend.h>
+
+namespace AZ::Dom
+{
+    Visitor::Result Backend::ReadFromStream(AZ::IO::GenericStream* stream, Visitor& visitor, size_t maxSize)
+    {
+        size_t length = stream->GetLength();
+        if (length > maxSize)
+        {
+            return AZ::Failure(VisitorError(VisitorErrorCode::InternalError, "Stream is too large."));
+        }
+        AZStd::string buffer;
+        buffer.resize(length);
+        stream->Read(length, buffer.data());
+        return ReadFromString(buffer, Lifetime::Temporary, visitor);
+    }
+
+    Visitor::Result Backend::ReadFromStringInPlace(AZStd::string& buffer, Visitor& visitor)
+    {
+        return ReadFromString(buffer, Lifetime::Persistent, visitor);
+    }
```

Comment:
This implementation doesn't work if the goal is to read from a `mutable` buffer.

The `ReadFromString` accepts an `AZStd::string_view`, therefore making the buffer argument immutable.

Also I feel like this implementation can be easily be replicated by making an `AZ::IO::ByteContainerStream<AZStd::string>` from the input string and then just invoking ReadFromStream.

I am not really a fan of special cases for AZStd::string, where there are equally viable buffer types that will be used(AZStd::fixed_string, AZStd::vector<char>).

Do you have some example of how a backend will use to avoid string allocations?

Because I feel like the `ReadFromStream` implementation should be the real implementation that the other functions call

Reply:
`ReadFromStringInPlace` is used by the `rapidjson` backend to fairly significant performance effect - this is just the "default" implementation that doesn't reap any of the performance benefits from specializing this method.

This could be adapted to take a (guaranteed null delimited) `char*` but I'd like to keep this path, as it is a pretty sizable throughput boost for the JSON case (see the `DomDeserializeToDocumentInPlace` numbers above)

Reply:
Do you have that `ReadFromStringInPlace` implementation for `rapidjson` as part of this PR or in a branch somewhere?

Because you can the same In Place performance with `ReadFromStream` as well by using an `AZ::IO::MemoryStream` that was passed a mutable `void*` buffer to which causes the `MemoryStreamMode::MSM_READWRITE` flag to be set

Looking at the [In Situ parsing](https://rapidjson.org/md_doc_dom.html#InSituParsing) page of rapidjson

Reply:
It's part of this PR, but I was fully unaware of `MSM_READWRITE` - if I can flag a stream as being writable this can actually be removed, that's pretty cool! I'll take a look

Reply:
OK yeah if you'd be cool with lightly amending `GenericStream`'s API with something like `void* TryGetWritableMemoryBuffer()` that I could use to pull out a buffer from a memory stream that would work for this (but I will need something like that to get the actual in-memory data unless RTTI is an option now).

Reply:
Well when I look at the AzStringStream from as a implementation of the rapidjson string API, I see no reason why it can't accept a char* and current size parameter.
That way rapidjson can use it in-situ parsing of any buffer, be that an `AZStd::vector<char>`, std::string, `AZStd::fixed_vector<char, Num>` or `AZStd::fixed_string<Num>`

Reply:
Think you maybe talking passed eachother. I think @lumberyard-employee-dm is advocating for a more generic input to a block of memory than `AZStd::string`, while @NicholasVanSickle is concerned about having to deal with streams.
Imho `AZ::IO::ByteContainerStream<AZStd::string>` can't be supported because what this function needs is an already fully loaded buffer. The function should however also not take an `AZStd::string` but a `char* buffer` and a `size_t bufferLength` for maximum flexibility so it can be called by a variety of buffers. Optionally wrapper functions can be added such as:
```
Visitor::Result Backend::ReadFromStringInPlace(AZStd::string& buffer, Visitor& visitor)
{
    return ReadFromStringInPlace(buffer.c_str(), buffer.size(), visitor);
}
```
although why add the noise?